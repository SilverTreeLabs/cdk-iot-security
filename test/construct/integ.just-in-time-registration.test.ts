import * as path from 'path';
import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import {
  RestApi,
  AuthorizationType,
  IAuthorizer,
} from '@aws-cdk/aws-apigateway';
import { Function, InlineCode, Runtime } from '@aws-cdk/aws-lambda';
import { Bucket } from '@aws-cdk/aws-s3';
import { App, Stack } from '@aws-cdk/core';
import {
  JustInTimeRegistration,
  LackOfAuthorizerError,
} from '../../src/just-in-time-registration';

test('CaRegisterApi integration test', () => {
  process.env.BASE_PATH = __dirname;
  process.env.APPS_PATH = path.resolve(__dirname, '..', '..', 'src', 'lambda-assets');
  const app = new App();
  const stack = new Stack(app, 'test-stack');
  const verifierStack = new Stack(app, 'verifier-stack');
  const name = 'test-case';
  const anotherStack = new Stack(app, 'another-stack');
  const bucket = new Bucket(anotherStack, 'userProvidedBucket');
  new JustInTimeRegistration(stack, name, {
    verifiers: [{
      name: 'test_verifier',
      lambdaFunction: new Function(verifierStack, name, {
        code: new InlineCode('exports.handler = () => { return true; }'),
        runtime: Runtime.NODEJS_12_X,
        handler: 'index.js',
      }),
    }],
    upload: {
      bucket: bucket,
      prefix: 'test',
    },
  });

  const expectedIds = {
    resource: {
      Ref: 'CaRegisterApitestcaseregister4C4AC1E7',
    },
    api: {
      Ref: 'CaRegisterApitestcaseFF571B6E',
    },
  };

  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();
  expect(stack).toHaveResourceLike('AWS::ApiGateway::Resource', {
    PathPart: 'register',
    RestApiId: expectedIds.api,
  });
  expect(stack).toHaveResourceLike('AWS::ApiGateway::Method', {
    HttpMethod: 'POST',
    RestApiId: expectedIds.api,
    ResourceId: expectedIds.resource,
  });
  expect(stack).toCountResources('AWS::Lambda::Function', 2);
  expect(stack).toCountResources('AWS::IAM::Role', 4);
  expect(stack).toHaveResourceLike('AWS::IAM::Role', {
    RoleName: 'CaRegistrationRoleName-' + name,
  });
  // expect(stack).toHaveResourceLike('AWS::IAM::Role', {
  //   RoleName: 'DeviceActivatorRoleName-' + name,
  // });
  expect(stack).toHaveResourceLike('AWS::IAM::Role', {
    RoleName: 'ReceptorPushRoleName-' + name,
  });
  expect(stack).toCountResources('AWS::SQS::Queue', 1);
});

test('CaRegisterApi integration test without specifying a verifier', () => {
  process.env.BASE_PATH = __dirname;
  process.env.APPS_PATH = path.resolve(__dirname, '..', '..', 'src', 'lambda-assets');
  const app = new App();
  const stack = new Stack(app, 'test-stack');
  const name = 'test-case';
  const anotherStack = new Stack(app, 'another-stack');
  const bucket = new Bucket(anotherStack, 'userProvidedBucket');
  new JustInTimeRegistration(stack, name, {
    upload: {
      bucket: bucket,
      prefix: 'test',
    },
  });
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();
});

test('CaRegisterApi integration test with a specified RestApi and the IAM Authrozation Type is specified', () => {
  process.env.BASE_PATH = __dirname;
  process.env.APPS_PATH = path.resolve(__dirname, '..', '..', 'src', 'lambda-assets');
  const app = new App();
  const stack = new Stack(app, 'test-stack');
  const name = 'test-case';
  const anotherStack = new Stack(app, 'another-stack');
  const bucket = new Bucket(anotherStack, 'userProvidedBucket');
  const providedRestApi = new RestApi(stack, 'test-api');
  new JustInTimeRegistration(stack, name, {
    restApiConfig: {
      restApi: providedRestApi,
      authorizationType: AuthorizationType.IAM,
    },
    upload: {
      bucket: bucket,
      prefix: 'test',
    },
  });
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();
});

test('CaRegisterApi integration test with a specified RestApi and the IAM Authrozation Type is specified', () => {
  process.env.BASE_PATH = __dirname;
  process.env.APPS_PATH = path.resolve(__dirname, '..', '..', 'src', 'lambda-assets');
  const app = new App();
  const stack = new Stack(app, 'test-stack');
  const name = 'test-case';
  const anotherStack = new Stack(app, 'another-stack');
  const bucket = new Bucket(anotherStack, 'userProvidedBucket');
  const providedRestApi = new RestApi(stack, 'test-api');
  new JustInTimeRegistration(stack, name, {
    restApiConfig: {
      restApi: providedRestApi,
      authorizationType: AuthorizationType.IAM,
    },
    upload: {
      bucket: bucket,
      prefix: 'test',
    },
  });
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();
});

test('Initialize CaRegisterApi with a Cognito-Authorized RestApi', () => {
  process.env.BASE_PATH = __dirname;
  process.env.APPS_PATH = path.resolve(__dirname, '..', '..', 'src', 'lambda-assets');
  const app = new App();
  const stack = new Stack(app, 'test-stack');
  const name = 'test-case';
  const anotherStack = new Stack(app, 'another-stack');
  const bucket = new Bucket(anotherStack, 'userProvidedBucket');
  const providedRestApi = new RestApi(stack, 'test-api');
  const providedIAuthorizer: IAuthorizer = {
    authorizerId: 'authorizer_id',
    authorizationType: AuthorizationType.COGNITO,
  };
  new JustInTimeRegistration(stack, name, {
    restApiConfig: {
      restApi: providedRestApi,
      authorizationType: AuthorizationType.COGNITO,
      authorizer: providedIAuthorizer,
    },
    upload: {
      bucket: bucket,
      prefix: 'test',
    },
  });
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();
});

test('Initialize CaRegisterApi with a Custom-Authorized RestApi', () => {
  process.env.BASE_PATH = __dirname;
  process.env.APPS_PATH = path.resolve(__dirname, '..', '..', 'src', 'lambda-assets');
  const app = new App();
  const stack = new Stack(app, 'test-stack');
  const name = 'test-case';
  const anotherStack = new Stack(app, 'another-stack');
  const bucket = new Bucket(anotherStack, 'userProvidedBucket');
  const providedRestApi = new RestApi(stack, 'test-api');
  const providedIAuthorizer: IAuthorizer = {
    authorizerId: 'authorizer_id',
    authorizationType: AuthorizationType.CUSTOM,
  };
  new JustInTimeRegistration(stack, name, {
    restApiConfig: {
      restApi: providedRestApi,
      authorizationType: AuthorizationType.CUSTOM,
      authorizer: providedIAuthorizer,
    },
    upload: {
      bucket: bucket,
      prefix: 'test',
    },
  });
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();
});

test('Initialize CaRegisterApi with a Cognito-Authorized RestApi but missing authorizer', () => {
  process.env.BASE_PATH = __dirname;
  process.env.APPS_PATH = path.resolve(__dirname, '..', '..', 'src', 'lambda-assets');
  const app = new App();
  const stack = new Stack(app, 'test-stack');
  const name = 'test-case';
  const anotherStack = new Stack(app, 'another-stack');
  const bucket = new Bucket(anotherStack, 'userProvidedBucket');
  const providedRestApi = new RestApi(stack, 'test-api');
  expect(() => {
    new JustInTimeRegistration(stack, name, {
      restApiConfig: {
        restApi: providedRestApi,
        authorizationType: AuthorizationType.COGNITO,
      },
      upload: {
        bucket: bucket,
        prefix: 'test',
      },
    });
  }).toThrowError(LackOfAuthorizerError);
});