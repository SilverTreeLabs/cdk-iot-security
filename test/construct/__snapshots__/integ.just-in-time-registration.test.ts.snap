// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaRegisterApi integration test 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters8b325ee7c020e93e73f9a41b2868268d1da3a8f9c9aa9a88326eb5bd48fc817dArtifactHashEE3A1FF1": Object {
      "Description": "Artifact hash for asset \\"8b325ee7c020e93e73f9a41b2868268d1da3a8f9c9aa9a88326eb5bd48fc817d\\"",
      "Type": "String",
    },
    "AssetParameters8b325ee7c020e93e73f9a41b2868268d1da3a8f9c9aa9a88326eb5bd48fc817dS3Bucket29041231": Object {
      "Description": "S3 bucket for asset \\"8b325ee7c020e93e73f9a41b2868268d1da3a8f9c9aa9a88326eb5bd48fc817d\\"",
      "Type": "String",
    },
    "AssetParameters8b325ee7c020e93e73f9a41b2868268d1da3a8f9c9aa9a88326eb5bd48fc817dS3VersionKey0FAE22DE": Object {
      "Description": "S3 key for asset version \\"8b325ee7c020e93e73f9a41b2868268d1da3a8f9c9aa9a88326eb5bd48fc817d\\"",
      "Type": "String",
    },
    "AssetParametersdce59a21d01c5c28ca59012c2b90d8a97c6e7eea001621c0981f733f972fddfdArtifactHash52761772": Object {
      "Description": "Artifact hash for asset \\"dce59a21d01c5c28ca59012c2b90d8a97c6e7eea001621c0981f733f972fddfd\\"",
      "Type": "String",
    },
    "AssetParametersdce59a21d01c5c28ca59012c2b90d8a97c6e7eea001621c0981f733f972fddfdS3BucketDED6A3ED": Object {
      "Description": "S3 bucket for asset \\"dce59a21d01c5c28ca59012c2b90d8a97c6e7eea001621c0981f733f972fddfd\\"",
      "Type": "String",
    },
    "AssetParametersdce59a21d01c5c28ca59012c2b90d8a97c6e7eea001621c0981f733f972fddfdS3VersionKeyB44852DD": Object {
      "Description": "S3 key for asset version \\"dce59a21d01c5c28ca59012c2b90d8a97c6e7eea001621c0981f733f972fddfd\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CaRegisterApitestcaseCaRegistrationFunctiontestcase2E1CEFE5": Object {
      "DependsOn": Array [
        "CaRegisterApitestcaseCaRegistrationFunctiontestcaseServiceRoleDefaultPolicy19C0B7B6",
        "CaRegisterApitestcaseCaRegistrationFunctiontestcaseServiceRole26C522A1",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters8b325ee7c020e93e73f9a41b2868268d1da3a8f9c9aa9a88326eb5bd48fc817dS3Bucket29041231",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8b325ee7c020e93e73f9a41b2868268d1da3a8f9c9aa9a88326eb5bd48fc817dS3VersionKey0FAE22DE",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8b325ee7c020e93e73f9a41b2868268d1da3a8f9c9aa9a88326eb5bd48fc817dS3VersionKey0FAE22DE",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "BUCKET_NAME": Object {
              "Fn::ImportValue": "another-stack:ExportsOutputRefuserProvidedBucket2349EF2D343936C6",
            },
            "BUCKET_PREFIX": "test",
            "DEIVCE_ACTIVATOR_QUEUE_URL": Object {
              "Ref": "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorQueuetestcase881E94AF",
            },
            "DEIVCE_ACTIVATOR_ROLE_ARN": Object {
              "Fn::GetAtt": Array [
                "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorQueuetestcaseDeviceActivatorQueuePushingRoletestcaseD04102E0",
                "Arn",
              ],
            },
            "test_verifier": Object {
              "Fn::ImportValue": "verifier-stack:ExportsOutputFnGetAtttestcase9BC7DE4DArn30A02383",
            },
          },
        },
        "Handler": "app.handler",
        "MemorySize": 256,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CaRegisterApitestcaseCaRegistrationFunctiontestcaseServiceRole26C522A1",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CaRegisterApitestcaseCaRegistrationFunctiontestcase372EACDE": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "iam:PassRole",
                "iot:RegisterCACertificate",
                "iot:GetRegistrationCode",
                "iot:CreateTopicRule",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CaRegisterApitestcaseCaRegistrationFunctiontestcase372EACDE",
        "Roles": Array [
          Object {
            "Ref": "CaRegisterApitestcaseCaRegistrationFunctiontestcaseServiceRole26C522A1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CaRegisterApitestcaseCaRegistrationFunctiontestcaseServiceRole26C522A1": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CaRegisterApitestcaseCaRegistrationFunctiontestcaseServiceRoleDefaultPolicy19C0B7B6": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::ImportValue": "another-stack:ExportsOutputFnGetAttuserProvidedBucket2349EF2DArnA89CCA27",
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::ImportValue": "another-stack:ExportsOutputFnGetAttuserProvidedBucket2349EF2DArnA89CCA27",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CaRegisterApitestcaseCaRegistrationFunctiontestcaseServiceRoleDefaultPolicy19C0B7B6",
        "Roles": Array [
          Object {
            "Ref": "CaRegisterApitestcaseCaRegistrationFunctiontestcaseServiceRole26C522A1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcase5FE1F333": Object {
      "DependsOn": Array [
        "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcaseServiceRoleDefaultPolicyC32FC9D0",
        "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcaseServiceRole78132BE2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersdce59a21d01c5c28ca59012c2b90d8a97c6e7eea001621c0981f733f972fddfdS3BucketDED6A3ED",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdce59a21d01c5c28ca59012c2b90d8a97c6e7eea001621c0981f733f972fddfdS3VersionKeyB44852DD",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdce59a21d01c5c28ca59012c2b90d8a97c6e7eea001621c0981f733f972fddfdS3VersionKeyB44852DD",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "app.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcaseServiceRole78132BE2",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcaseDeviceActivationFunctionPolicytestcase71060DAE": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "iot:UpdateCertificate",
                "iot:DescribeCertificate",
                "lambda:InvokeFunction",
                "lambda:InvokeAsync",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcaseDeviceActivationFunctionPolicytestcase71060DAE",
        "Roles": Array [
          Object {
            "Ref": "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcaseServiceRole78132BE2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcaseServiceRole78132BE2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcaseServiceRoleDefaultPolicyC32FC9D0": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorQueuetestcase881E94AF",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcaseServiceRoleDefaultPolicyC32FC9D0",
        "Roles": Array [
          Object {
            "Ref": "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcaseServiceRole78132BE2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcaseSqsEventSourceteststackCaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorQueuetestcase67AD1AB0EC898E70": Object {
      "Properties": Object {
        "BatchSize": 1,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorQueuetestcase881E94AF",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorFunctiontestcase5FE1F333",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorQueuetestcase881E94AF": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorQueuetestcaseDeviceActivatorQueuePushingRoletestcaseD04102E0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "iot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "sqs:SendMessageBatch",
                    "sqs:SendMessage",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "CaRegisterApitestcaseDeviceActivatortestcaseDeviceActivatorQueuetestcase881E94AF",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "SqsPushPolicy",
          },
        ],
        "RoleName": "DeviceActivatorQueuePushingRoleName-test-case",
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
